### PR TITLE
Issues/83 sliding menu

### PR DESCRIPTION
**Pull request by aerych**
Refs #83

This PR is the first step at implementing the app's production UI.  There are lots of changes!  The PR has over 6800 additions, but fear not! Probably ~90% of that is in project boilerplate and the acknowledgements.html file (which can be ignored for now).

### What's new?

**SidebarContainerViewController**: a Swift port of the similar ObjC sliding menu control from Simplenote iOS. This version comes with a couple of modifications, most notably is a UIView that overlays the main view and intercepts taps when the menu is visible. This version of the control also listens for notifications to toggle the sidebar so child controllers do not need direct knowledge of the container. This (mostly?) gets rid of the need for a delegate, at least as far as this implementation goes but I've kept the delegate code in the hopes we can later extract this take on the class to share between apps.

**MenuViewController**: This is the controller that makes up the sidebar menu. It shows profile information, available sites, links to an about screen, and has a log out button.

**AboutViewController**: Provides a screen to consolidate things like app version, and links to terms, privacy, etc.  It is modeled after a similar view in WordPress iOS.

**WebViewController**: SFSafarViewController does not load local files, so I cobbled together a custom web viewer based off of a similar class in WordPress iOS. This version is drastically simplified (e.g. no need for auth and other things).

**Appearance.swift**: A convenience class to centralize applying styles to different views and controls. Think of it as a Newspack analogue to WPStyleGuide.

There are some other odds and ends like the UserView, CircularImageView, and a convenience SimpleTableViewCell class.

### What's the status?

The new sliding menu is enabled and can be opened from the Folders list.  Please pay no mind to the Folders list, it's next up for a facelift ;)

The menu, about, and web controllers should all properly support dark mode and be reasonably well styled.  A final design pass for extra polish with be made in the future.  For now this should be good to be getting along with.

### Known issues

- The Newspack icon added in this PR is the wrong size, and I didn't have the patience to figure out how to recreate a universal pdf that is the correct size.  I'll sort this out in the future with a proper design assist.  

- I think it would be nice to have a vertical rule to create a nice clean edge between the menu and main views. The best place is probably the MenuViewController but it would need to be rebuilt as a regular UIViewController (not a UITableViewController) so tableView != view in order to properly support an overlay.  This PR is large enough already so I'll deal with that another day. 

- Privacy Policy, Terms, and Acknowledgements should be viewed as placeholders.

- There is currently no progress or offine handling in the WebViewController.  This PR is large so some things needed to be deferred.

@jleandroperez are you game for a review? I know there's a lot here but I promise it will go fast! 

### Testing!
Now for the fun part 😁 
**Launch screen**
- [x] Launch the app.  Confirm you see a (some what fuzzy but still) epic new logo on the launch screen.  Confirm there is a smooth transition to the main view.
**Sidebar**
- [x] Tap (or swipe) to open the side bar.
- [x] Tap (or swipe) on the sliver of the main view to close the sidebar.
- [x] Confirm that taps on the main screen are intercepted by the overlay while the side bar is visible. You should not be able to, say, tap a cell to view assets, or edit a title while the menu is open.
**Menu**
- [x] Confirm your avatar loads, and you see your user name.
- [x] Confirm your site is listed.
- [x] Confirm you can log out.
- [x] Confirm the About item opens the About screen modally.
**About**
- [x] Confirm you see a (some what fuzzy but still) epic Newspack logo.
- [x] Confirm the Newspack item launches the web view and loads the Newspack website modally.
- [x] Confirm the Privacy item launches the web view and loads Automattic's privacy policy
- [x] Confirm the Terms item launches the web view and loads the WordPress terms page.
- [x] Confirm the Source Code item launches the web view and loads the Newspack github project.
- [x] Confirm the Acknowledgements item launches the web view and loads the acknowledgements.html file.
- [x] Confirm you can close the About screen via the cross button or by a swipe down.
**WebViewController**
- [x] Test the reload button. Confirm the page reloads.
- [x] Test the globe button to open the page in safari.
- [x] Test the share button. Confirm a share sheet appears.
- [x] Try clicking a link on a web page. Then, after the new page loads, test the back and forward buttons.
- [x] Confirm that the page title and domain are updated when changing pages.
- [x] Confirm you can close the web view via the cross button or by a swipe down.
**Bonus**
- [x] Perform a smoke test on an iPad. It should mostly work. ;)

### Examples

Simulator Screen Shot - iPhone SE (2nd generation) - 2020-07-22 at 16 38 43
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-07-22 at 16 38 46
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-07-22 at 16 38 54

### CODE REVIEW

File: Newspack/Newspack/Controllers/AboutViewController.swift [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    func configureButtons() {
+        closeButton.image = UIImage.gridicon(.cross)
+    }
+
+    func configureHeader() {
+        var frame = headerView.frame
+        frame.size.height = 132.0
+        headerView.frame = frame
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/AboutViewController.swift]: You can just set the height directly:

```
headerView.frame.size.height = 132
```

(Plus: constant maybe?)

File: Newspack/Newspack/Controllers/AboutViewController.swift [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        return sections.count
+    }
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return sections[section].rows.count
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: SimpleTableViewCell.reuseIdentifier, for: indexPath)
+        Appearance.style(cell: cell)
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/AboutViewController.swift:61]: Nice!

File: Newspack/Newspack/Controllers/MainNavController.swift [hunk 3]
```diff
@@ -26,16 +26,18 @@ final class MainNavController: UINavigationController {
...
         guard viewControllers.first is InitialViewController else {
             return
         }
 
-        let controller = MainStoryboard.instantiateViewController(withIdentifier: .siteMenu)
+        let menuController = MainStoryboard.instantiateViewController(withIdentifier: .menu)
+        let folderController = MainStoryboard.instantiateViewController(withIdentifier: .folderList)
+        let navController = UINavigationController(rootViewController: folderController)
+
+        let controller = SidebarContainerViewController(mainViewController: navController, sidebarViewController: menuController)
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/MainNavController.swift:38]: This. Right there. This is SO beautiful!! (along with the MainStoryboard init!!)

File: Newspack/Newspack/Controllers/MenuViewController.swift [hunk 4]
```diff
@@ -0,0 +1,191 @@
+import UIKit
+import Gridicons
+
+class MenuViewController: UITableViewController {
+
+    @IBOutlet var userView: UserView!
+
+    var receipt: Any?
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/MenuViewController.swift]: Private maybe?

File: Newspack/Newspack/Controllers/MenuViewController.swift [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        guard let details = StoreContainer.shared.accountStore.currentAccount?.details else {
+            return
+        }
+        userView.configure(with: details.username, gravatar: details.avatarURL)
+    }
+
+    func configureInsets() {
+        var insets = tableView.contentInset
+        insets.top = 44.0
+        tableView.contentInset = insets
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/MenuViewController.swift]: (Same note as above!) no need to extract the insets, you can hit directly the property 🔥

File: Newspack/Newspack/Controllers/SidebarContainerViewController.swift [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+        static let sidebarWidth = CGFloat(300)
+        static let sidebarAnimationThreshold = CGFloat(0.15)
+        static let sidebarAnimationDuration = 0.35
+        static let sidebarAnimationDamping = CGFloat(10)
+        static let sidebarAnimationInitialVelocity = CGVector(dx: -10, dy: 0)
+        static let sidebarAnimationCompletionMin = CGFloat(0.001)
+        static let sidebarAnimationCompletionMax = CGFloat(0.999)
+        static let sidebarAnimationCompletionFactorFull = CGFloat(1.0)
+        static let sidebarAnimationCompletionFactorZero = CGFloat(0.0)
+    }
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/SidebarContainerViewController.swift:32]: This looks SO much better in Swift!!

File: Newspack/Newspack/Controllers/SidebarContainerViewController.swift [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+        view.addGestureRecognizer(panGestureRecognizer)
+
+        // The following order is important. The mainView should be lowest.
+        // The capture view should be second so it covers the main view.
+        // The sidebar view should be topmost.
+        view.addSubview(mainView)
+        attachCaptureView()
+        attachSidebarView()
+
+        NotificationCenter.default.addObserver(self, selector: #selector(handleToggleSidebarNotification(notification:)), name: SidebarContainerViewController.toggleSidebarNotification, object: nil)
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/SidebarContainerViewController.swift:147]: I really really like the idea of cutting ties between all the things. May borrow your work later sir!! ❤️

**Review by jleandroperez — APPROVED** [review]
LOVED this PR, it's **very** solid, end to end.
Code looks sharp, all test cases passed without a glitch.

Nice work Eric!!

:shipit:

**Comment by aerych** [comment]
Thank you @jleandroperez!!

**Pull request merged**